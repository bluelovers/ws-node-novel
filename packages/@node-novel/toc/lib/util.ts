/**
 * Created by user on 2018/11/14/014.
 */

import { _trim, createSortCallback, defaultSortCallback, EnumToLowerCase } from '@node-novel/sort';
import { array_unique } from 'array-hyper-unique';
import FastGlob = require('fast-glob');
import fs = require('fs-iconv');
import { IMdconfMeta, mdconf_parse } from 'node-novel-info';
import BluebirdPromise = require('bluebird');
import StrUtil = require('str-util');

export async function loadReadmeMeta<T extends IMdconfMeta = IMdconfMeta>(file: string): Promise<T>
{
	return fs.readFile(file)
		.then(function (data)
		{
			return mdconf_parse(data, {
				// 當沒有包含必要的內容時不產生錯誤
				throw: false,
				// 允許不標準的 info 內容
				lowCheckLevel: true,
			});
		})
		.catch(function ()
		{
			return null;
		})
		;
}

export function loadReadmeMetaSync<T extends IMdconfMeta = IMdconfMeta>(file: string): T
{
	try
	{
		let data = fs.readFileSync(file);

		// @ts-ignore
		return mdconf_parse(data, {
			// 當沒有包含必要的內容時不產生錯誤
			throw: false,
			// 允許不標準的 info 內容
			lowCheckLevel: true,
		})
	}
	catch (e)
	{

	}

	return null;
}

export function getNovelTitles<T extends IMdconfMeta = IMdconfMeta>(meta: T): string[]
{
	if (meta && meta.novel)
	{
		let arr = [
				'title',
				'title_source',
				'title_jp',
				'title_ja',
				'title_zh',
				'title_tw',
				'title_cn',
			].concat(Object.keys(meta.novel))
			.reduce(function (a, key: string)
			{
				if (key.indexOf('title') === 0)
				{
					a.push(meta.novel[key])
				}

				return a
			}, [])
		;

		if (meta.novel.series)
		{
			arr.push(meta.novel.series.name);
			arr.push(meta.novel.series.name_short);
		}

		arr = array_unique(arr.filter(v => v && ![
			'undefined',
			'長編 【連載】',
			'連載中',
		].includes(v)));

		return arr;
	}

	return [];
}

export function globFirst(...argv: Parameters<typeof FastGlob["stream"]>): BluebirdPromise<string>
{
	return new BluebirdPromise<string>(function (resolve, reject)
	{
		let fgs = FastGlob.stream(...argv);

		fgs.on('data', (entry) =>
		{
			resolve(entry);

			// @ts-ignore
			fgs.destroy();
		});
		fgs.once('error', reject);
		fgs.once('end', () => resolve(undefined));
	})
}

export function md_href(href: string)
{
	return href.split('/').map(encodeURIComponent).join('/');
}

export function md_anchor_gitee(title: string)
{
	let anchor = title
		.replace(/[a-z]+/ig, function (s)
		{
			return s.toLowerCase();
		})
		.replace(/[\.．\/／　＠@（）\(\)～~]/g, '')
		.replace(/[ ]/g, '-')
	;

	return md_href(anchor);
}

export function md_link_escape(text: string)
{
	return text.replace(/[\[\]]/g, function (s)
	{
		return '\\' + s;
	})
}

export const tocSortCallback = createSortCallback({
	dotNum: true,
	transpileBase(input: string, isSub?: any)
	{
		let s = StrUtil.toHalfWidth(input);
		return s
	},
	toLowerCase: EnumToLowerCase.toLocaleLowerCase,
});
