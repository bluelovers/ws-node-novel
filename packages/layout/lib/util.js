"use strict";
/**
 * Created by user on 2019/5/29.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("./index");
const blank_line_1 = require("blank-line");
function _isIwordsArray(value) {
    return Array.isArray(value) && (value.length > 1);
}
exports._isIwordsArray = _isIwordsArray;
function _isIwordsArray2(value) {
    return Array.isArray(value) && value.length == 1 && typeof value[0] == 'function';
}
exports._isIwordsArray2 = _isIwordsArray2;
function _isIwordsUserSp(value) {
    return typeof value.s == 'string' && new RegExp(`${index_1.SP_KEY}(.+)$`).test(value.s);
}
exports._isIwordsUserSp = _isIwordsUserSp;
function _handleTextLayout(html, options) {
    if (!html.match(/[^\n]\n[^\n]/g)) {
        let [min, mid, max] = blank_line_1.default(html);
        if (min > 2) {
            options.allow_lf2 = false;
        }
        if (max >= 3) {
            if (min > 2) {
                let r = new RegExp(`\\n{${min - 1}}(\\n+)`, 'g');
                html = html
                    //.replace(/\n{2}(\n*)/g, '$1')
                    .replace(r, '$1');
            }
            html = html
                .replace(/\n{3,}/g, "\n\n\n");
        }
        //console.log(options);
        if (!options.allow_lf2) {
            html = html
                .replace(/\n\n/g, "\n");
        }
    }
    /*
    html = html
    // for ts
        .toString()
        .replace(/([^\n「」【】《》“”『』（）\[\]"](?:[！？?!。]*)?)\n((?:[—]+)?[「」“”【】《》（）『』])/ug, "$1\n\n$2")

        .replace(/([「」【】《》“”『』（）―\[\]"](?:[！？?!。]*)?)\n((?:\u3000*)[^\n「」“”【】《》（）『』])/ug, "$1\n\n$2")

        .replace(/([^\n「」【】《》“”『』（）\[\]"≪≫](?:[！？?!。]*)?)\n((?:[—]+)?[≪≫「」“”【】《》（）『』])/ug, "$1\n\n$2")

        .replace(/(）(?:[！？?!。]*)?)\n([「」【】《》『』“”])/ug, "$1\n\n$2")
    ;
    */
    html = html
        .replace(/^\n+|[\s\u3000]+$/g, '')
        .replace(/(\n){4,}/g, "\n\n\n\n" /* LF4 */);
    if (options.allow_lf3) {
        html = html
            .replace(/(\n){3,}/g, "\n\n\n" /* LF3 */);
    }
    else {
        html = html
            .replace(/(\n){3}/g, "\n\n" /* LF2 */);
    }
    return html;
}
exports._handleTextLayout = _handleTextLayout;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidXRpbC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbInV0aWwudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IjtBQUFBOztHQUVHOztBQUdILG1DQUFpQztBQUNqQywyQ0FBc0M7QUFFdEMsU0FBZ0IsY0FBYyxDQUFDLEtBQUs7SUFFbkMsT0FBTyxLQUFLLENBQUMsT0FBTyxDQUFDLEtBQUssQ0FBQyxJQUFJLENBQUMsS0FBSyxDQUFDLE1BQU0sR0FBRyxDQUFDLENBQUMsQ0FBQztBQUNuRCxDQUFDO0FBSEQsd0NBR0M7QUFFRCxTQUFnQixlQUFlLENBQUMsS0FBSztJQUVwQyxPQUFPLEtBQUssQ0FBQyxPQUFPLENBQUMsS0FBSyxDQUFDLElBQUksS0FBSyxDQUFDLE1BQU0sSUFBSSxDQUFDLElBQUksT0FBTyxLQUFLLENBQUMsQ0FBQyxDQUFDLElBQUksVUFBVSxDQUFDO0FBQ25GLENBQUM7QUFIRCwwQ0FHQztBQUVELFNBQWdCLGVBQWUsQ0FBQyxLQUFLO0lBRXBDLE9BQU8sT0FBTyxLQUFLLENBQUMsQ0FBQyxJQUFJLFFBQVEsSUFBSSxJQUFJLE1BQU0sQ0FBQyxHQUFHLGNBQU0sT0FBTyxDQUFDLENBQUMsSUFBSSxDQUFDLEtBQUssQ0FBQyxDQUFDLENBQUMsQ0FBQztBQUNqRixDQUFDO0FBSEQsMENBR0M7QUFFRCxTQUFnQixpQkFBaUIsQ0FBQyxJQUFZLEVBQUUsT0FBMkI7SUFFMUUsSUFBSSxDQUFDLElBQUksQ0FBQyxLQUFLLENBQUMsZUFBZSxDQUFDLEVBQ2hDO1FBQ0MsSUFBSSxDQUFDLEdBQUcsRUFBRSxHQUFHLEVBQUUsR0FBRyxDQUFDLEdBQUcsb0JBQVksQ0FBQyxJQUFJLENBQUMsQ0FBQztRQUV6QyxJQUFJLEdBQUcsR0FBRyxDQUFDLEVBQ1g7WUFDQyxPQUFPLENBQUMsU0FBUyxHQUFHLEtBQUssQ0FBQztTQUMxQjtRQUVELElBQUksR0FBRyxJQUFJLENBQUMsRUFDWjtZQUNDLElBQUksR0FBRyxHQUFHLENBQUMsRUFDWDtnQkFDQyxJQUFJLENBQUMsR0FBRyxJQUFJLE1BQU0sQ0FBQyxPQUFPLEdBQUcsR0FBRyxDQUFDLFNBQVMsRUFBRSxHQUFHLENBQUMsQ0FBQztnQkFFakQsSUFBSSxHQUFHLElBQUk7b0JBQ1gsK0JBQStCO3FCQUM3QixPQUFPLENBQUMsQ0FBQyxFQUFFLElBQUksQ0FBQyxDQUNqQjthQUNEO1lBRUQsSUFBSSxHQUFHLElBQUk7aUJBQ1QsT0FBTyxDQUFDLFNBQVMsRUFBRSxRQUFRLENBQUMsQ0FFN0I7U0FDRDtRQUVELHVCQUF1QjtRQUV2QixJQUFJLENBQUMsT0FBTyxDQUFDLFNBQVMsRUFDdEI7WUFDQyxJQUFJLEdBQUcsSUFBSTtpQkFDVCxPQUFPLENBQUMsT0FBTyxFQUFFLElBQUksQ0FBQyxDQUN2QjtTQUNEO0tBQ0Q7SUFFRDs7Ozs7Ozs7Ozs7O01BWUU7SUFFRixJQUFJLEdBQUcsSUFBSTtTQUNULE9BQU8sQ0FBQyxvQkFBb0IsRUFBRSxFQUFFLENBQUM7U0FDakMsT0FBTyxDQUFDLFdBQVcsdUJBQWEsQ0FDakM7SUFFRCxJQUFJLE9BQU8sQ0FBQyxTQUFTLEVBQ3JCO1FBQ0MsSUFBSSxHQUFHLElBQUk7YUFDVCxPQUFPLENBQUMsV0FBVyxxQkFBYSxDQUNqQztLQUNEO1NBRUQ7UUFDQyxJQUFJLEdBQUcsSUFBSTthQUNULE9BQU8sQ0FBQyxVQUFVLG1CQUFhLENBQ2hDO0tBQ0Q7SUFFRCxPQUFPLElBQUksQ0FBQztBQUNiLENBQUM7QUF4RUQsOENBd0VDIiwic291cmNlc0NvbnRlbnQiOlsiLyoqXG4gKiBDcmVhdGVkIGJ5IHVzZXIgb24gMjAxOS81LzI5LlxuICovXG5cbmltcG9ydCB7IEVudW1MRiwgSVRleHRMYXlvdXRPcHRpb25zLCBJV29yZHNBcnJheSwgSVdvcmRzQXJyYXkyLCBJV29yZHNVc2VyU1AgfSBmcm9tICcuL3R5cGVzJztcbmltcG9ydCB7IFNQX0tFWSB9IGZyb20gJy4vaW5kZXgnO1xuaW1wb3J0IGdldE1pbk1pZE1heCBmcm9tICdibGFuay1saW5lJztcblxuZXhwb3J0IGZ1bmN0aW9uIF9pc0l3b3Jkc0FycmF5KHZhbHVlKTogdmFsdWUgaXMgSVdvcmRzQXJyYXlcbntcblx0cmV0dXJuIEFycmF5LmlzQXJyYXkodmFsdWUpICYmICh2YWx1ZS5sZW5ndGggPiAxKTtcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIF9pc0l3b3Jkc0FycmF5Mih2YWx1ZSk6IHZhbHVlIGlzIElXb3Jkc0FycmF5Mlxue1xuXHRyZXR1cm4gQXJyYXkuaXNBcnJheSh2YWx1ZSkgJiYgdmFsdWUubGVuZ3RoID09IDEgJiYgdHlwZW9mIHZhbHVlWzBdID09ICdmdW5jdGlvbic7XG59XG5cbmV4cG9ydCBmdW5jdGlvbiBfaXNJd29yZHNVc2VyU3AodmFsdWUpOiB2YWx1ZSBpcyBJV29yZHNVc2VyU1Bcbntcblx0cmV0dXJuIHR5cGVvZiB2YWx1ZS5zID09ICdzdHJpbmcnICYmIG5ldyBSZWdFeHAoYCR7U1BfS0VZfSguKykkYCkudGVzdCh2YWx1ZS5zKTtcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIF9oYW5kbGVUZXh0TGF5b3V0KGh0bWw6IHN0cmluZywgb3B0aW9uczogSVRleHRMYXlvdXRPcHRpb25zKTogc3RyaW5nXG57XG5cdGlmICghaHRtbC5tYXRjaCgvW15cXG5dXFxuW15cXG5dL2cpKVxuXHR7XG5cdFx0bGV0IFttaW4sIG1pZCwgbWF4XSA9IGdldE1pbk1pZE1heChodG1sKTtcblxuXHRcdGlmIChtaW4gPiAyKVxuXHRcdHtcblx0XHRcdG9wdGlvbnMuYWxsb3dfbGYyID0gZmFsc2U7XG5cdFx0fVxuXG5cdFx0aWYgKG1heCA+PSAzKVxuXHRcdHtcblx0XHRcdGlmIChtaW4gPiAyKVxuXHRcdFx0e1xuXHRcdFx0XHRsZXQgciA9IG5ldyBSZWdFeHAoYFxcXFxueyR7bWluIC0gMX19KFxcXFxuKylgLCAnZycpO1xuXG5cdFx0XHRcdGh0bWwgPSBodG1sXG5cdFx0XHRcdC8vLnJlcGxhY2UoL1xcbnsyfShcXG4qKS9nLCAnJDEnKVxuXHRcdFx0XHRcdC5yZXBsYWNlKHIsICckMScpXG5cdFx0XHRcdDtcblx0XHRcdH1cblxuXHRcdFx0aHRtbCA9IGh0bWxcblx0XHRcdFx0LnJlcGxhY2UoL1xcbnszLH0vZywgXCJcXG5cXG5cXG5cIilcblx0XHRcdC8vLnJlcGxhY2UoL1xcbnsyfS9nLCBcIlxcblwiKVxuXHRcdFx0O1xuXHRcdH1cblxuXHRcdC8vY29uc29sZS5sb2cob3B0aW9ucyk7XG5cblx0XHRpZiAoIW9wdGlvbnMuYWxsb3dfbGYyKVxuXHRcdHtcblx0XHRcdGh0bWwgPSBodG1sXG5cdFx0XHRcdC5yZXBsYWNlKC9cXG5cXG4vZywgXCJcXG5cIilcblx0XHRcdDtcblx0XHR9XG5cdH1cblxuXHQvKlxuXHRodG1sID0gaHRtbFxuXHQvLyBmb3IgdHNcblx0XHQudG9TdHJpbmcoKVxuXHRcdC5yZXBsYWNlKC8oW15cXG7jgIzjgI3jgJDjgJHjgIrjgIvigJzigJ3jgI7jgI/vvIjvvIlcXFtcXF1cIl0oPzpb77yB77yfPyHjgIJdKik/KVxcbigoPzpb4oCUXSspP1vjgIzjgI3igJzigJ3jgJDjgJHjgIrjgIvvvIjvvInjgI7jgI9dKS91ZywgXCIkMVxcblxcbiQyXCIpXG5cblx0XHQucmVwbGFjZSgvKFvjgIzjgI3jgJDjgJHjgIrjgIvigJzigJ3jgI7jgI/vvIjvvInigJVcXFtcXF1cIl0oPzpb77yB77yfPyHjgIJdKik/KVxcbigoPzpcXHUzMDAwKilbXlxcbuOAjOOAjeKAnOKAneOAkOOAkeOAiuOAi++8iO+8ieOAjuOAj10pL3VnLCBcIiQxXFxuXFxuJDJcIilcblxuXHRcdC5yZXBsYWNlKC8oW15cXG7jgIzjgI3jgJDjgJHjgIrjgIvigJzigJ3jgI7jgI/vvIjvvIlcXFtcXF1cIuKJquKJq10oPzpb77yB77yfPyHjgIJdKik/KVxcbigoPzpb4oCUXSspP1viiariiavjgIzjgI3igJzigJ3jgJDjgJHjgIrjgIvvvIjvvInjgI7jgI9dKS91ZywgXCIkMVxcblxcbiQyXCIpXG5cblx0XHQucmVwbGFjZSgvKO+8iSg/OlvvvIHvvJ8/IeOAgl0qKT8pXFxuKFvjgIzjgI3jgJDjgJHjgIrjgIvjgI7jgI/igJzigJ1dKS91ZywgXCIkMVxcblxcbiQyXCIpXG5cdDtcblx0Ki9cblxuXHRodG1sID0gaHRtbFxuXHRcdC5yZXBsYWNlKC9eXFxuK3xbXFxzXFx1MzAwMF0rJC9nLCAnJylcblx0XHQucmVwbGFjZSgvKFxcbil7NCx9L2csIEVudW1MRi5MRjQpXG5cdDtcblxuXHRpZiAob3B0aW9ucy5hbGxvd19sZjMpXG5cdHtcblx0XHRodG1sID0gaHRtbFxuXHRcdFx0LnJlcGxhY2UoLyhcXG4pezMsfS9nLCBFbnVtTEYuTEYzKVxuXHRcdDtcblx0fVxuXHRlbHNlXG5cdHtcblx0XHRodG1sID0gaHRtbFxuXHRcdFx0LnJlcGxhY2UoLyhcXG4pezN9L2csIEVudW1MRi5MRjIpXG5cdFx0O1xuXHR9XG5cblx0cmV0dXJuIGh0bWw7XG59XG4iXX0=