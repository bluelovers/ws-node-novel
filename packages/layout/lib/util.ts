/**
 * Created by user on 2019/5/29.
 */

import { EnumLF, ITextLayoutOptions, IWordsArray, IWordsArray2, IWordsUserSP } from './types';
import { SP_KEY } from './index';
import getMinMidMax from 'blank-line';

export function _isIwordsArray(value): value is IWordsArray
{
	return Array.isArray(value) && (value.length > 1);
}

export function _isIwordsArray2(value): value is IWordsArray2
{
	return Array.isArray(value) && value.length == 1 && typeof value[0] == 'function';
}

export function _isIwordsUserSp(value): value is IWordsUserSP
{
	return typeof value.s == 'string' && new RegExp(`${SP_KEY}(.+)$`).test(value.s);
}

export function _handleTextLayout(html: string, options: ITextLayoutOptions): string
{
	if (!html.match(/[^\n]\n[^\n]/g))
	{
		let [min, mid, max] = getMinMidMax(html);

		if (min > 2)
		{
			options.allow_lf2 = false;
		}

		if (max >= 3)
		{
			if (min > 2)
			{
				let r = new RegExp(`\\n{${min - 1}}(\\n+)`, 'g');

				html = html
				//.replace(/\n{2}(\n*)/g, '$1')
					.replace(r, '$1')
				;
			}

			html = html
				.replace(/\n{3,}/g, "\n\n\n")
			//.replace(/\n{2}/g, "\n")
			;
		}

		//console.log(options);

		if (!options.allow_lf2)
		{
			html = html
				.replace(/\n\n/g, "\n")
			;
		}
	}

	/*
	html = html
	// for ts
		.toString()
		.replace(/([^\n「」【】《》“”『』（）\[\]"](?:[！？?!。]*)?)\n((?:[—]+)?[「」“”【】《》（）『』])/ug, "$1\n\n$2")

		.replace(/([「」【】《》“”『』（）―\[\]"](?:[！？?!。]*)?)\n((?:\u3000*)[^\n「」“”【】《》（）『』])/ug, "$1\n\n$2")

		.replace(/([^\n「」【】《》“”『』（）\[\]"≪≫](?:[！？?!。]*)?)\n((?:[—]+)?[≪≫「」“”【】《》（）『』])/ug, "$1\n\n$2")

		.replace(/(）(?:[！？?!。]*)?)\n([「」【】《》『』“”])/ug, "$1\n\n$2")
	;
	*/

	html = html
		.replace(/^\n+|[\s\u3000]+$/g, '')
		.replace(/(\n){4,}/g, EnumLF.LF4)
	;

	if (options.allow_lf3)
	{
		html = html
			.replace(/(\n){3,}/g, EnumLF.LF3)
		;
	}
	else
	{
		html = html
			.replace(/(\n){3}/g, EnumLF.LF2)
		;
	}

	return html;
}
