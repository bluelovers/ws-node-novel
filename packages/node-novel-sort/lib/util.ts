/**
 * Created by user on 2020/6/5.
 */

import { ITriggerData } from './types';

export function _match(a: string, b: string, {
	r,
	mainFn,
}: ITriggerData)
{
	let ta: RegExpExecArray;
	let tb: RegExpExecArray;

	if ((ta = r.exec(a)) && (tb = r.exec(b)))
	{
		let r = parseFloat(ta[0]) - parseFloat(tb[0]);

		if (r !== 0)
		{
			return r;
		}

		let a1 = ta.input.slice(ta[0].length);
		let b1 = tb.input.slice(tb[0].length);

		if (a1 != b1)
		{
			let i = 0;

			while (typeof a1[i] != 'undefined' && a1[i] === b1[i] && (!/^\d$/.test(b1[i])))
			{
				i++;
			}

			a1 = a1.slice(i);
			b1 = b1.slice(i);
		}

		return mainFn(a1, b1, true);
	}
}

export function _trim(input: string): string
{
	return input
		.replace(/^[_\s]+(\d+)/, '$1')
		.replace(/^\D(\d+)/, '$1')
		.trim()
		;
}
