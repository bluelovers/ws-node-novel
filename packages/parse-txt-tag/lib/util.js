"use strict";
/**
 * Created by user on 2020/1/4.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const fullhalf_1 = require("str-util/lib/fullhalf");
const tags_1 = require("./tags");
function _fixRubyInnerContext(innerContext) {
    let fn = _replaceHtmlTag(($0, $1, $2, $3) => {
        return `<${$1}${$2}>${$3}</${$1}>`;
    });
    return innerContext
        .replace(tags_1.reHtmlRubyRt, fn)
        .replace(tags_1.reHtmlRubyRp, fn);
}
exports._fixRubyInnerContext = _fixRubyInnerContext;
function _replaceHtmlTag(replacer) {
    return ($0, $1, $2, ...argv) => {
        $1 = fullhalf_1.toHalfWidth($1);
        $2 = fullhalf_1.toHalfWidth($2);
        return replacer($0, $1, $2, ...argv);
    };
}
exports._replaceHtmlTag = _replaceHtmlTag;
function _convertHtmlTag001(input) {
    return input
        .replace(new RegExp("&lt;|\\u003C|\uFF1C" /* OPEN */, 'ig'), '<')
        .replace(new RegExp("&gt;|\\u003E|\uFF1E" /* CLOSE */, 'ig'), '>');
}
exports._convertHtmlTag001 = _convertHtmlTag001;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoidXRpbC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbInV0aWwudHMiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IjtBQUFBOztHQUVHOztBQUVILG9EQUFvRDtBQUNwRCxpQ0FBaUU7QUFFakUsU0FBZ0Isb0JBQW9CLENBQUMsWUFBb0I7SUFFeEQsSUFBSSxFQUFFLEdBQUcsZUFBZSxDQUFDLENBQUMsRUFBRSxFQUFFLEVBQUUsRUFBRSxFQUFFLEVBQUUsRUFBRSxFQUFFLEVBQUU7UUFFM0MsT0FBTyxJQUFJLEVBQUUsR0FBRyxFQUFFLElBQUksRUFBRSxLQUFLLEVBQUUsR0FBRyxDQUFBO0lBQ25DLENBQUMsQ0FBQyxDQUFDO0lBRUgsT0FBTyxZQUFZO1NBQ2pCLE9BQU8sQ0FBQyxtQkFBWSxFQUFFLEVBQUUsQ0FBQztTQUN6QixPQUFPLENBQUMsbUJBQVksRUFBRSxFQUFFLENBQUMsQ0FDekI7QUFDSCxDQUFDO0FBWEQsb0RBV0M7QUFFRCxTQUFnQixlQUFlLENBQUMsUUFBNEQ7SUFFM0YsT0FBTyxDQUFDLEVBQVUsRUFBRSxFQUFVLEVBQUUsRUFBVSxFQUFFLEdBQUcsSUFBYyxFQUFFLEVBQUU7UUFHaEUsRUFBRSxHQUFHLHNCQUFXLENBQUMsRUFBRSxDQUFDLENBQUM7UUFDckIsRUFBRSxHQUFHLHNCQUFXLENBQUMsRUFBRSxDQUFDLENBQUM7UUFFckIsT0FBTyxRQUFRLENBQUMsRUFBRSxFQUFFLEVBQUUsRUFBRSxFQUFFLEVBQUUsR0FBRyxJQUFJLENBQUMsQ0FBQTtJQUNyQyxDQUFDLENBQUE7QUFDRixDQUFDO0FBVkQsMENBVUM7QUFFRCxTQUFnQixrQkFBa0IsQ0FBQyxLQUFhO0lBRS9DLE9BQU8sS0FBSztTQUNWLE9BQU8sQ0FBQyxJQUFJLE1BQU0sbUNBQW1CLElBQUksQ0FBQyxFQUFFLEdBQUcsQ0FBQztTQUNoRCxPQUFPLENBQUMsSUFBSSxNQUFNLG9DQUFvQixJQUFJLENBQUMsRUFBRSxHQUFHLENBQUMsQ0FDakQ7QUFDSCxDQUFDO0FBTkQsZ0RBTUMiLCJzb3VyY2VzQ29udGVudCI6WyIvKipcbiAqIENyZWF0ZWQgYnkgdXNlciBvbiAyMDIwLzEvNC5cbiAqL1xuXG5pbXBvcnQgeyB0b0hhbGZXaWR0aCB9IGZyb20gJ3N0ci11dGlsL2xpYi9mdWxsaGFsZic7XG5pbXBvcnQgeyByZUh0bWxSdWJ5UnQsIHJlSHRtbFJ1YnlScCwgRW51bUh0bWxUYWcgfSBmcm9tICcuL3RhZ3MnO1xuXG5leHBvcnQgZnVuY3Rpb24gX2ZpeFJ1YnlJbm5lckNvbnRleHQoaW5uZXJDb250ZXh0OiBzdHJpbmcpXG57XG5cdGxldCBmbiA9IF9yZXBsYWNlSHRtbFRhZygoJDAsICQxLCAkMiwgJDMpID0+XG5cdHtcblx0XHRyZXR1cm4gYDwkeyQxfSR7JDJ9PiR7JDN9PC8keyQxfT5gXG5cdH0pO1xuXG5cdHJldHVybiBpbm5lckNvbnRleHRcblx0XHQucmVwbGFjZShyZUh0bWxSdWJ5UnQsIGZuKVxuXHRcdC5yZXBsYWNlKHJlSHRtbFJ1YnlScCwgZm4pXG5cdFx0O1xufVxuXG5leHBvcnQgZnVuY3Rpb24gX3JlcGxhY2VIdG1sVGFnKHJlcGxhY2VyOiAoKHN1YnN0cmluZzogc3RyaW5nLCAuLi5hcmdzOiBzdHJpbmdbXSkgPT4gc3RyaW5nKSlcbntcblx0cmV0dXJuICgkMDogc3RyaW5nLCAkMTogc3RyaW5nLCAkMjogc3RyaW5nLCAuLi5hcmd2OiBzdHJpbmdbXSkgPT5cblx0e1xuXG5cdFx0JDEgPSB0b0hhbGZXaWR0aCgkMSk7XG5cdFx0JDIgPSB0b0hhbGZXaWR0aCgkMik7XG5cblx0XHRyZXR1cm4gcmVwbGFjZXIoJDAsICQxLCAkMiwgLi4uYXJndilcblx0fVxufVxuXG5leHBvcnQgZnVuY3Rpb24gX2NvbnZlcnRIdG1sVGFnMDAxKGlucHV0OiBzdHJpbmcpXG57XG5cdHJldHVybiBpbnB1dFxuXHRcdC5yZXBsYWNlKG5ldyBSZWdFeHAoRW51bUh0bWxUYWcuT1BFTiwgJ2lnJyksICc8Jylcblx0XHQucmVwbGFjZShuZXcgUmVnRXhwKEVudW1IdG1sVGFnLkNMT1NFLCAnaWcnKSwgJz4nKVxuXHRcdDtcbn1cbiJdfQ==