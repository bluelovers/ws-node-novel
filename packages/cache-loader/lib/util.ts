/**
 * Created by user on 2019/1/6/006.
 */

import moment from 'moment';
import { defaultSortCallback, naturalCompare } from '@node-novel/sort';
import { toHalfWidth } from '@lazy-cjk/fullhalf';
import sortObject from 'sort-object-keys2';
import { _trim } from '@node-novel/sort/lib/util';
import { EnumToLowerCase } from '@node-novel/sort/lib/types';
import { createSortCallback } from '@node-novel/sort/lib/core';

let defaultOffset = 8;

export { naturalCompare }

export function createMoment(...argv)
{
	return moment(...argv).utcOffset(defaultOffset);
}

const todayMoment = createMoment().startOf('day');

export let todayMomentTimestamp = todayMoment.valueOf();
export let todayMomentOffset = todayMoment.utcOffset();

export default todayMomentTimestamp

export function getTodayMomentTimestamp()
{
	return todayMoment.valueOf();
}

export function refreshTodayMoment()
{
	let k = createMoment()
		//.add(7, 'days')
		.startOf('day')
	;

	if (k.valueOf() != todayMoment.valueOf())
	{
		todayMoment.set(k.toObject());

		return true;
	}
}

Object.defineProperties(exports, {
	todayMomentTimestamp: {
		get: getTodayMomentTimestamp,
	},
	todayMomentOffset: {
		get()
		{
			return todayMoment.utcOffset();
		},
	},
});

export const cacheSortCallback = createSortCallback({
	dotNum: true,
	transpileBase(input: string, isSub?: any)
	{
		let s = toHalfWidth(input);
		return s
	},
	toLowerCase: EnumToLowerCase.toLocaleLowerCase,
});

export function freezeProperty<T>(who: T, prop: keyof T, freeze?: boolean)
{
	if (freeze)
	{
		try
		{
			// @ts-ignore
			who[prop] = Object.freeze(who[prop]);
		}
		catch (e)
		{

		}
	}

	Object.defineProperty(who, prop, {
		configurable: false,
		writable: false,
	});

	return who;
}

export function baseSortObject<T>(data: T): T
{
	return sortObject(data, {
		useSource: true,
		keys: Object.keys(data).sort(),
	});
}

export function tryRequireFS(): typeof import('fs-extra')
{
	let fs: typeof import('fs-extra');

	try
	{
		fs = require('fs-extra');
		return fs;
	}
	catch (e)
	{

	}
}

export function parsePathMainBase(pathMain: string)
{
	let is_out: boolean = null;
	let pathMain_base: string = undefined;
	let pathMain_out: string = undefined;

	if (pathMain != null)
	{
		let _m = pathMain.match(/^(.+?)(_out)?$/);

		is_out = !!_m[2];
		pathMain_base = _m[1];
		pathMain_out = pathMain_base + '_out';
	}

	return {
		is_out,
		pathMain_base,
		pathMain_out,
	}
}
