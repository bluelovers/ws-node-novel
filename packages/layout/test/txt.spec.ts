import testCaseList from './data/simple';
import textLayout from '../index';
import { inspect } from 'util';

describe(`textLayout.textlayout`, () =>
{
	testCaseList
		.forEach((testCase) => {
			test(inspect(testCase[1]), () =>
			{

				let actual = textLayout.textlayout(testCase[0]);

				let expected = testCase[1];

				if (typeof expected === 'string')
				{
					expect(actual).toContain(expected);
				}
				else if (typeof expected === 'function')
				{
					expect(expected(actual, testCase)).toBeTruthy();
				}
				else
				{
					expect(actual).toMatch(expected);
				}

				expect(actual).toMatchSnapshot();

			});
		})
	;

})
