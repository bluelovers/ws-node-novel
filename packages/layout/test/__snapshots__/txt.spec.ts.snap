// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textLayout.textlayout /」\\n統/ 1`] = `
"心情愉快的諾艾爾稍微動了一下之後張開雙眼。
頓時臉頰紅潤著。

諾艾爾「…那個、早上好。統夜先生」
統夜「嗯、早安。諾艾爾」

雖說諾艾爾仍舊眷戀著他的身旁，但還是爬起來了。
原本隔著她輕薄粉色的睡衣傳來的體溫也離去了。"
`;
